File: src/main/java/gui/MainWindow.java
package gui;

import controller.Draw;
import controller.MainWindowController;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.SimpleDoubleProperty;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.control.Button;
import javafx.scene.control.ColorPicker;
import javafx.scene.control.TextInputDialog;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.stage.Stage;
import model.enums.BasicForm;

public class MainWindow {
    private MainWindowController controller;

    // TODO: Guardar as constantes em outro lugar
    private final String title = "Bora Desenhar";
    private BasicForm formaAtual = BasicForm.POINT;
    private int drawDiameter = 4;
    private int panePadding = 10;

    private Color primaryColor = Color.BLACK;
    private Color secondaryColor = Color.WHITE;

    private DoubleProperty canvasWidth;
    private DoubleProperty canvasHeight;
    private DoubleProperty diffHeight;

    private void iniciateObservables() {
        canvasWidth = new SimpleDoubleProperty(0.0);
        canvasHeight = new SimpleDoubleProperty(0.0);
        diffHeight = new SimpleDoubleProperty(0.0);
    }

    public MainWindow(MainWindowController controller, Stage stage) {
        iniciateObservables();

        this.controller = controller;
        iniciateWindow(stage);
    }

    public void iniciateWindow(Stage stage) {
        // define titulo da janela
        stage.setTitle(title);

        // define largura e altura da janela
        stage.setWidth(500);
        stage.setHeight(500);

        // Painel para os componentes
        VBox pane = new VBox(panePadding);

        // componente para desenho
        Canvas canvas = new Canvas();
//        controller.context = canvas.getGraphicsContext2D();

        // componente para desenhar graficos

        GraphicsContext gc = canvas.getGraphicsContext2D();

        // componente para os botões
        HBox buttons = new HBox(5);
        includeButtons(buttons, gc);
        diffHeight.bind(buttons.heightProperty().add(panePadding));

        // Eventos de mouse
        canvas.setOnMouseMoved(event -> controller.updateWindowTitleWithCoordinates(stage, event));
        canvas.setOnMouseExited(event -> controller.updateWindowTitleWithoutCoordinates(stage));
        canvas.setOnMousePressed(event -> controller.drawForm(gc, event));

        // atributos do painel
        pane.setBackground(new Background(new BackgroundFill(Color.WHITESMOKE, CornerRadii.EMPTY, Insets.EMPTY)));
        pane.getChildren().addAll(buttons, canvas); // posiciona o componente de desenho
        letResizeCanvas(pane, canvas);

        // cria e insere cena
        Scene scene = new Scene(pane);
        stage.setScene(scene);
        stage.show();
    }

    private void letResizeCanvas(Pane pane, Canvas canvas) {
        // inicializa variaveis observables
        canvas.widthProperty().bindBidirectional(canvasWidth);
        canvas.heightProperty().bindBidirectional(canvasHeight);

        // Conecta o tamanho do canvas ao tamanho do painel
        canvas.widthProperty().bind(
                pane.widthProperty());
        canvas.heightProperty().bind(
                pane.heightProperty().subtract(diffHeight));
    }

    private void includeButtons(HBox buttons, GraphicsContext gc) {
        // desenhar pontos
        Button point = new Button("Ponto");
        point.setOnAction(click -> controller.selectPoint());

        // desenhar linhas
        Button line = new Button("Reta");
        line.setOnAction(click -> controller.selectLine());

        // desenhar circulos
        Button circle = new Button("Círculo");
        circle.setOnAction(click -> controller.selectCircle());

        // desenhar outras formas
        Button drawForm = new Button("Outras Formas");
        drawForm.setOnAction(click -> drawFormDialog(gc)); // TODO: passar OnActionListener para o controller

        // limpar canvas
        Button clear = new Button("Limpar");
        clear.setOnAction(click -> clear(gc) ); // TODO: passar OnActionListener para o controller


        // color choices
        ColorPicker primaryColorPicker = new ColorPicker();
        ColorPicker secondaryColorPicker = new ColorPicker();
        controller.bindColors(primaryColorPicker, secondaryColorPicker);

        VBox colors = new VBox(5);
        colors.getChildren().addAll(primaryColorPicker, secondaryColorPicker);
        colors.setMinHeight(60);

        buttons.getChildren().addAll(point, line, circle, drawForm, clear, colors);
    }

    private void clear(GraphicsContext gc) {
        gc.clearRect(0, 0, canvasWidth.get(), canvasHeight.get());
        controller.resetDraw();
    }

    private void drawFormDialog(GraphicsContext gc) {
        TextInputDialog drawFormDialog = new TextInputDialog();

        drawFormDialog.setTitle("Desenhar Forma");
        drawFormDialog.setHeaderText("Insira o número de divisões");
        drawFormDialog.setContentText("Divisões:");

        drawFormDialog.showAndWait()
                .filter(input -> controller.isIntGreaterThanZero(input))
                .ifPresent(div -> {
                    clear(gc);
                    Draw.Handler.drawLineForm(
                            gc,
                            canvasWidth.get(),
                            canvasHeight.get(),
                            Integer.parseInt(div),
                            1,
                            primaryColor);
                });
    }

}

File: src/main/kotlin/view/MainWindow.kt
package view

import controller.MainWindowController
import javafx.geometry.Insets
import javafx.scene.Scene
import javafx.scene.canvas.Canvas
import javafx.scene.canvas.GraphicsContext
import javafx.scene.control.*
import javafx.scene.image.ImageView
import javafx.scene.input.KeyCombination
import javafx.scene.layout.*
import javafx.stage.Stage
import model.constants.Constants
import model.enums.BasicForm

class MainWindow(private val controller: MainWindowController, stage: Stage) {

    init {
        iniciateWindow(stage)
    }

    private fun iniciateWindow(stage: Stage) {
        // componente para desenho
        val canvas = Canvas().apply {
            width = Constants.CANVAS_WIDTH
            height = Constants.CANVAS_HEIGHT
        }
        val supportCanvas = Canvas().apply {
            width = Constants.CANVAS_WIDTH
            height = Constants.CANVAS_HEIGHT
        }

        // componente para desenhar graficos
        val context = canvas.graphicsContext2D

        // componente para os botões
        val buttons = includeButtons(context)

        val canvasPane = Pane().apply {
            border = Border(BorderStroke(
                    Constants.DEFAULT_SEPARATOR_COLOR,
                    BorderStrokeStyle.SOLID,
                    CornerRadii.EMPTY,
                    BorderWidths(0.0, 0.0, 0.0, 2.0)
            ))
            background = Background(BackgroundFill(
                    Constants.DEFAULT_BACKGROUND_COLOR,
                    CornerRadii.EMPTY,
                    Insets.EMPTY
            ))
            children.addAll(canvas, supportCanvas)
            supportCanvas.toFront()
        }
        controller.mainCanvas = canvas
        controller.supportCanvas = supportCanvas

        // Eventos de mouse
        handleMouseEvents(canvasPane, stage, context)

        val miniWindow = BorderPane().apply {
            setPrefSize(200.0, 200.0)
            border = Border(BorderStroke(
                    Constants.DEFAULT_SEPARATOR_COLOR,
                    BorderStrokeStyle.SOLID,
                    CornerRadii.EMPTY,
                    BorderWidths(2.0)
            ))

            center = ImageView().apply {
                fitWidth = 200.0
                fitHeight = 200.0
                imageProperty().bind(controller.canvasSnapshot)
                isPreserveRatio = true
            }
        }

        val menuBar = includeMenu(stage)

        // Painel para os componentes
        val pane = GridPane().apply {
            rowConstraints.addAll(
                    RowConstraints(),
                    RowConstraints().apply {
                        vgrow = Priority.ALWAYS
                    },
                    RowConstraints()
            )
            columnConstraints.addAll(
                    ColumnConstraints(),
                    ColumnConstraints().apply {
                        hgrow = Priority.ALWAYS
                    }
            )

            add(buttons, 0, 1)
            add(miniWindow, 0, 2)
            add(canvasPane, 1, 1, 1 , 2)
            canvasPane.also {
                val row = GridPane.getRowIndex(it)
                val column = GridPane.getColumnIndex(it)
                GridPane.setMargin(it, Insets(0.0, 0.0, 0.0, controller.panePadding))

                it.maxHeightProperty().bind(rowConstraints[row].maxHeightProperty())
                it.maxWidthProperty().bind(columnConstraints[column].maxWidthProperty())
            }
            miniWindow.also {
                GridPane.setMargin(it, Insets(0.0, 0.0, controller.panePadding, controller.panePadding))
            }


            add(menuBar, 0, 0, 2, 1)
        }

        // cria e insere cena
        val scene = Scene(pane)
        controller.bindCanvasSize(scene)

        // define e inicia o stage
        stage.run {
            // define titulo da janela
            title = controller.title

            // define largura e altura da janela
            width = 800.0
            height = 700.0

            this.scene = scene
            setShortcuts(this, context)
            show()
        }
    }

    private fun includeButtons(context: GraphicsContext) : Pane {
        // Painel para as cores
        val colors = VBox(5.0).apply {
            // color choices
            val primaryColorPicker = ColorPicker()
            val secondaryColorPicker = ColorPicker()
            controller.bindColors(
                    primaryColorPicker.valueProperty(),
                    secondaryColorPicker.valueProperty()
            )
            children.addAll(primaryColorPicker, secondaryColorPicker)
        }

        // Painel para os botões
        return VBox(5.0).apply {
            padding = Insets(10.0)
            minWidth = 200.0
            children.addAll(
                    colors,
                    ComboBox<BasicForm>().apply {
                        itemsProperty().bind(controller.availableForms)
                        valueProperty().bindBidirectional(controller.selectedForm)

                        setOnAction { value?.let { controller.selectForm(it) } }
                    },
                    Button().apply { // desenhar forma com linhas
                        text = "Curva com retas"
                        disableProperty().bind(controller.clippingActive)
                        setOnAction { controller.selectLineForm(context) }
                    },
                    Button().apply { // limpar canvas
                        text = "Limpar"
                        setOnAction { controller.clear(context) }
                    },
                    Button().apply { // desfazer ação
                        text = "Desfazer"
                        setOnAction { controller.undo(context) }
                    },
                    Button().apply { // refazer ação
                        text = "Refazer"
                        setOnAction { controller.redo(context) }
                    },
                    CheckBox().apply { // clipping do canvas
                        text = "Clip"
                        selectedProperty().bindBidirectional(controller.clipping)
                        selectedProperty().addListener { _ -> controller.clip() }
                    }
            )
        }
    }

    private fun includeMenu(stage: Stage): MenuBar {
        return MenuBar().apply {
            menus.addAll(
                    Menu("Arquivo").apply {
                        items.addAll(
                                MenuItem("Abrir (Ctrl + O)").apply {
                                    setOnAction { controller.openFigureFromFile(stage) }
                                },
                                MenuItem("Salvar (Ctrl + S)").apply {
                                    setOnAction { controller.saveFigureOnFile(stage) }
                                }
                        )
                    }
            )
        }
    }

    private fun handleMouseEvents(canvas: Pane, stage: Stage, context: GraphicsContext) {
        canvas.setOnMouseMoved { event -> controller.updateWindowTitleWithCoordinates(stage, event) }
        canvas.setOnMouseExited { controller.updateWindowTitleWithoutCoordinates(stage) }
        canvas.setOnMousePressed { event -> controller.drawForm(context, event) }
    }

    private fun setShortcuts(stage: Stage, context: GraphicsContext) {
        val shortcuts = mapOf(
                KeyCombination.valueOf("Ctrl+O") to Runnable { controller.openFigureFromFile(stage) },
                KeyCombination.valueOf("Ctrl+S") to Runnable { controller.saveFigureOnFile(stage) },
                KeyCombination.valueOf("Ctrl+Z") to Runnable { controller.undo(context) },
                KeyCombination.valueOf("Ctrl+Y") to Runnable { controller.redo(context) },
                KeyCombination.valueOf("Ctrl+Shift+Z") to Runnable { controller.redo(context) },
                KeyCombination.valueOf("Ctrl+Delete") to Runnable { controller.clear(context) }
        )

        stage.scene.accelerators.putAll(shortcuts)
    }
}
